### DISCUSSION TITLE
Nginx Proxy Manager and 2FA

### DISCUSSION THREAD

**Discussion by ulysserevient** [dani-garcia/vaultwarden#4672] [Q&A] [answered]
Hi,
we are using VaultWarden under Nginx Proxy Manager, and it works correctly, unless we activate 2FA

With 2FA, each time someone try to connect, a red popup saying "An unexpected error occurred" appears.

In the browser console : 
![image](https://github.com/dani-garcia/vaultwarden/assets/11133694/0b49b8d9-662d-49d5-8527-cfcba38e4fbe)

In the server log: 

```
[2024-06-25 10:58:47.185][request][INFO] POST /identity/accounts/prelogin
[2024-06-25 10:58:47.186][response][INFO] (prelogin) POST /identity/accounts/prelogin => 200 OK
[2024-06-25 10:58:47.607][request][INFO] POST /identity/connect/token
[2024-06-25 10:58:47.680][error][ERROR] 2FA token not provided
[2024-06-25 10:58:47.680][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
[2024-06-25 10:58:55.866][request][INFO] POST /identity/accounts/prelogin
[2024-06-25 10:58:55.866][response][INFO] (prelogin) POST /identity/accounts/prelogin => 200 OK
[2024-06-25 10:58:56.277][request][INFO] POST /identity/connect/token
[2024-06-25 10:58:56.353][error][ERROR] 2FA token not provided
[2024-06-25 10:58:56.353][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
```

Don't know to repair that !
Is there a custom config to put in NPM ?

**Comment by That-Guy-Jack** [comment]
Just had the same issue here not sure whats going on.

**Comment by kennysamuerto** [comment]
Hi, I have the same problem with NPM and VaultWarden. When I activate the 2FA, the error is reproduced preventing to enter the code.

**Comment by ulysserevient** [comment]
Hi,
At this time, I don't have find any solution !

**Comment by melo-afk** [comment]
I just had the same issue and I think the cause for it was custom error pages that I added in ngnix proxy manager. 

https://github.com/dani-garcia/vaultwarden/issues/4870#issuecomment-2296299370 states that issues occur when the reverse proxy catches the error and returns its own version

To check if this is the cause, go to your browser’s network tab and attempt to log in. If the 400 response includes your custom error page instead of the expected JSON response, then this is likely the issue.

To use the default error page again you need to add the following snippet into the advanced config

```  
# lock down the admin page / delete this part if you need to access it
location /admin {
  	return 404;
  }

location / {
    proxy_pass          $forward_scheme://$server:$port;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Host $host;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";


    # Disable custom error page
    error_page 404 = /404_default;
    location = /404_default {
        internal;
        # => default NGINX 404 page
    }
}
```
Yes, the logs always say error 400 bad request but the npm config only needs the default 404 page and it does not work when you change 404 to 400.

hope this fixes you problem as well

**Reply by maidlinux01** [reply]
I would like to thank you, this one fixed a similar issue I had.
I went over updating and recreating the container several times, and doing a full backup, and it never worked.
Turns out some time ago I just installed custom error pages at my `nginx proxy manager` container, I never thought that could cause this issue on `vaultwarden`

## Screenshot of the debug steps, we can see that the response is the custom `html`  error page instead of `json`
![image](https://github.com/user-attachments/assets/5c98a10a-f75a-4833-891b-25b2f531bcc3)

## Screnshot of the custom confing at `npm`
![image](https://github.com/user-attachments/assets/3b508e60-2b61-4070-9f94-e356fdf011da)

## Screenshot of the expected `json` response
![image](https://github.com/user-attachments/assets/03325d25-a19d-42da-a98c-31fecef58643)

**Comment by ulysserevient** [comment, marked as answer]
Ho !!!
Thank you very much, it solves our problem!

Here is my NPM conf :

```
location /admin {
  return 404;
  }

location / {
  proxy_http_version 1.1;

  proxy_set_header Host $host;
  proxy_set_header X-Real-IP $remote_addr;
  proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
  proxy_set_header X-Forwarded-Proto $scheme;

  # Disable custom error page
  error_page 404 = /404_default;
  location = /404_default {
	internal;
	# => default NGINX 404 page
  }

  proxy_pass http://X.X.X.X:80;
  }
  
location /notifications {
  proxy_http_version 1.1;
  proxy_set_header Upgrade $http_upgrade;
  proxy_set_header Connection $http_connection;

  proxy_set_header Host $host;
  proxy_set_header X-Real-IP $remote_addr;
  proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
  proxy_set_header X-Forwarded-Proto $scheme;

  proxy_pass http://X.X.X.X:80;
  }
```
